### PR TITLE
refactor(hive): rework hive configuration

### PR DESCRIPTION
**Pull request by gboutry**
#### Additional comments:

This PR separates hiveserver2 from metastore configuration.

This PR templates a "new" hive service: `metaSchemaTool`. It's used to replace `schemaTool` since `schemaTool` cannot read configuration from `metastore-site.xml`

This PR introduces a :warning: breaking change in variables. Connection information moved from the dict `hive_site` to `metastore_site`. (Only the metastore needs them)

For example, this breaks in tdp-getting-started, need to update these variables later:

```diff
-hive_site:
+metastore_site:
   javax.jdo.option.ConnectionURL: '<your value>'
   javax.jdo.option.ConnectionDriverName: '<you driver>'
```

This PR removes the command `beeline_auto`, and instead, templates `beeline-site.xml` with a default connection string. If no connection argument is passed to `beeline`, the default connection string is used.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: roles/hive/common/templates/hive-env.sh.j2 [hunk 1]
```diff
@@ -53,16 +53,15 @@ if [ "$SERVICE" = "metastore" ]; then
   export JMX_OPTS="{{ jmx_common_opts }} -Dcom.sun.management.jmxremote.password.file={{ hive_ms_conf_dir }}/jmxremote.password  {{ jmx_exporter_hms_opts }} "
 
   #export HADOOP_HEAPSIZE= # Setting for HiveMetastore
-  export HADOOP_OPTS="$HADOOP_OPTS $JMX_OPTS -Xloggc:{{ hive_log_dir }}/metastore-gc-%t.log -XX:+UseG1GC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCCause -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=10M -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath={{ hive_log_dir }}/hms_heapdump.hprof -Dhive.log.dir={{ hive_log_dir }} -Dhive.log.file=metastore.log"
+  export HADOOP_OPTS="$HADOOP_OPTS $JMX_OPTS -Xloggc:{{ hive_log_dir }}/metastore-gc-%t.log -XX:+UseG1GC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCCause -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=10M -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath={{ hive_log_dir }}/hms_heapdump.hprof -Djava.security.auth.login.config={{ hive_ms_conf_dir }}/jaas.conf -Djavax.security.auth.useSubjectCredsOnly=false"
```
Comment (rpignolet) [inline roles/hive/common/templates/hive-env.sh.j2:56]: Why using `hive_ms_conf_dir`, a metastore variable inside hive common template ? Hiveserver2 tasks do not template this file ?

Comment (gboutry) [reply roles/hive/common/templates/hive-env.sh.j2:56]: I kept kind of the same behavior as before, because this file is templated by both `hiveserver2` and `metastore` tasks. And the variables are available inside `hive.yml`.

There are 3 possibilities:
- Keep as it is
- Add a `hive_conf_dir` var, and at the task level plug the right variable
- Break the templates symlink and keep template files separated by component

Comment (rpignolet) [reply roles/hive/common/templates/hive-env.sh.j2:56]: The code is inside `if [ "$SERVICE" = "metastore" ]; then` so it is ok.

**Review by rpignolet — APPROVED** [review]

**Review by gboutry — COMMENTED** [review]
After more advanced testing, there's an issue with impersonation on sql ddl queries, while there's no issue on the dql queries. Looking into it.

**Review by mehdibn — APPROVED** [review]

File: roles/hive/common/templates/hive-server2.service.j2 [hunk 2]
```diff
@@ -6,11 +6,11 @@ Wants=network-online.target
 [Service]
 User={{ hive_user }}
 Group={{ hadoop_group }}
-PIDFile={{ hive_pid_dir }}/hiveserver2.pid
 ExecStart={{ hive_install_dir }}/bin/start-hiveserver2.sh
 SuccessExitStatus=143
 LimitNOFILE=64000
 Restart={{ hiveserver2_restart }}
+SyslogIdentifier=hive-server2
```
Comment (leopaul36) [inline roles/hive/common/templates/hive-server2.service.j2:13]: Why is this needed?

**Comment by leopaul36** [comment]
# Tests

## Beeline

As `tdp_user`:
```
ERROR : FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. MetaException(message:Got exception: org.apache.hadoop.security.AccessControlException Permission denied: user=hive, access=EXECUTE, inode="/user/tdp_user":tdp_user:tdp_user:drwx------
```

Impersonation fails HS2 seems to have the right parameters:

```
0: jdbc:hive2://master-01.tdp:2181,master-02.> set hive.server2.enable.doAs;
+--------------------------------+
|              set               |
+--------------------------------+
| hive.server2.enable.doAs=true  |
+--------------------------------+
```

In the `metastore.log`, we can see that it is `hive/master-02.tdp` and not `tdp_user` making the request:

```
2022-10-10T16:51:23,302  INFO [pool-11-thread-10] metastore.HiveMetaStore: 10: get_databases: @hive#
2022-10-10T16:51:23,302  INFO [pool-11-thread-10] HiveMetaStore.audit: ugi=hive/master-02.tdp@REALM.TDP	ip=192.168.56.12	cmd=get_databases: @hive#
```

`hiveserver2.log` tells us that it's opening a metastore client with `ugi=tdp_user`... 

```
[HiveServer2-HttpHandler-Pool: Thread-342] metastore.RetryingMetaStoreClient: RetryingMetaStoreClient proxy=class org.apache.hadoop.hive.ql.metadata.SessionHiveMetaStoreClient ugi=tdp_user (auth:PROXY) via hive/master-02.tdp@REALM.TDP (auth:KERBEROS) retries=1 delay=1 lifetime=0
```

## Spark 3 Shell

With the following code:

```
val sqlContext = new org.apache.spark.sql.hive.HiveContext(sc);
sqlContext.sql("show databases").show();
```

In the `metastore.log`, we can see that it is indeed `tdp_user` making the request:

```
2022-10-10T16:50:03,526  INFO [pool-11-thread-13] metastore.HiveMetaStore: 14: get_databases: @hive#*
2022-10-10T16:50:03,526  INFO [pool-11-thread-13] HiveMetaStore.audit: ugi=tdp_user@REALM.TDP	ip=192.168.56.10	cmd=get_databases: @hive#*
```

## HIVE-20776

From what I understand in the discussions at [HIVE-20776](https://issues.apache.org/jira/browse/HIVE-20776), there is a difference in the way HS2 and Spark sets the proxy user which could be responsible for our issue:

```
HMS uses the hadoop proxy user system to know the end-user, no? This is how Spark interacts with HMS. Spark sets the hadoop proxy user and HMS knows the end-user through that. Why not use the same mechanism from HS2?
```

I tried to cherry-pick from [HIVE-20776](https://github.com/apache/hive/commit/dfd63d97902b359e1643e955a4d070ac983debd5) and had a few conflicts.

But then Hive Standalone Metastore won't compile...

image

**Comment by gboutry** [comment]
Nice catch @mehdibn, I took a glance at Hive's source code, and the `CLIService` (hive server2's service), is still depending on the deprecated property `hive.metastore.sasl.enabled`.

I will rebase on master.

**Comment by mehdibn** [comment]
> Nice catch @mehdibn, I took a glance at Hive's source code, and the `CLIService` (hive server2's service), is still depending on the deprecated property `hive.metastore.sasl.enabled`.
> 
> I will rebase on master.

Only your PR is nice @gboutry :) unfortunately, we will cross many contradictions in this sort of projects ..

**Comment by gboutry** [comment]
Force push is removing jaas configuration added by the rework as it is no longer necessary

**Review by leopaul36 — APPROVED** [review]
Works great good job @mehdibn @gboutry !
